### PR TITLE
Candid prop tests

### PR DESCRIPTION
**Pull request by bdemann**

### CODE REVIEW

File: tests/property/ic_api/candid/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import { candidCompiler, candidDecode, candidEncode, IDL, query } from 'azle';
+
+export default class {
+    @query([IDL.Text], IDL.Text)
+    candidCompilerQuery(candidPath: string): string {
```
Comment (lastmjs) [inline tests/property/ic_api/candid/src/index.ts:5]: Git rid of candidCompiler entirely

File: tests/property/ic_api/candid/test/test.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import { getCanisterId } from 'azle/dfx';
+import { runTests } from 'azle/test';
+
+import { createActor } from './dfx_generated/canister';
+import { getTests } from './tests';
+
+const canister = createActor(getCanisterId('canister'), {
```
Comment (lastmjs) [inline tests/property/ic_api/candid/test/test.ts:7]: Use getCanisterActor

File: tests/property/ic_api/candid/tsconfig.json
Comment (lastmjs) [inline tests/property/ic_api/candid/tsconfig.json:1]: top-level await

File: tests/property/ic_api/candid/test/tests.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
+import { ActorSubclass } from '@dfinity/agent';
+import { defaultPropTestParams, expect, it, Test } from 'azle/test';
+import fc from 'fast-check';
+import { TextDecoder, TextEncoder } from 'util';
+
+// @ts-ignore this path may not exist when these tests are imported into other test projects
```
Comment (lastmjs) [inline tests/property/ic_api/candid/test/tests.ts:6]: Remove comment if not necessary

File: tests/property/ic_api/candid/test/tests.ts
Comment (lastmjs) [inline tests/property/ic_api/candid/test/tests.ts:1]: We need to create arbitrary Candid strings

File: tests/property/ic_api/candid/test/tests.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+                        expect(encodedBytes).toEqual(candidBytes);
+                    }
+                ),
+                defaultPropTestParams
+            );
+        });
+
+        // TODO candidCompilerQuery is not yet implemented
+        it.skip('should compile Candid correctly', async () => {
```
Comment (lastmjs) [inline tests/property/ic_api/candid/test/tests.ts:64]: Remove this

**Review by lastmjs — CHANGES_REQUESTED** [review]

File: tests/property/ic_api/candid/src/index.ts (hunk 1 above)
Comment (bdemann) [reply tests/property/ic_api/candid/src/index.ts]: done

File: tests/property/ic_api/candid/test/test.ts (hunk 2 above)
Comment (bdemann) [reply tests/property/ic_api/candid/test/test.ts]: done

File: tests/property/ic_api/candid/tsconfig.json
Comment (bdemann) [reply tests/property/ic_api/candid/tsconfig.json:1]: wont work because of jest

File: tests/property/ic_api/candid/test/tests.ts (hunk 3 above)
Comment (bdemann) [reply tests/property/ic_api/candid/test/tests.ts]: done

File: tests/property/ic_api/candid/test/tests.ts [hunk 5]
```diff
@@ -1,110 +1,78 @@
-import { ActorSubclass } from '@dfinity/agent';
-import { defaultPropTestParams, expect, it, Test } from 'azle/test';
+globalThis._azleExperimental = true;
+import { execSyncPretty } from 'azle/src/build/stable/utils/exec_sync_pretty';
```
Comment (lastmjs) [inline tests/property/ic_api/candid/test/tests.ts:2]: leave a linebreak

Comment (lastmjs) [inline tests/property/ic_api/candid/test/tests.ts:2]: execSyncPretty should be used everywhere in this file right?

File: tests/property/ic_api/candid/test/tests.ts [hunk 6]
```diff
@@ -1,110 +1,78 @@
...
-                defaultPropTestParams
+        please('install didc', async () => {
+            execSync(
+                `cargo install --git https://github.com/dfinity/candid --rev 5d3c7c35da652d145171bc071ac11c63d73bf803 --force didc`,
+                { stdio: 'inherit' }
             );
         });
 
-        it('should decode Candid bytes and then encode correctly, with UTF-8 verification', async () => {
+        it('should encode and decode candid values correctly', async () => {
```
Comment (lastmjs) [inline tests/property/ic_api/candid/test/tests.ts:28]: Let's add a comment into this file explaining how to make the tests better, explaining the limitations, explaining that we could add arbitrary candid string creation to our candid arbitraries

File: tests/property/ic_api/candid/test/tests.ts [hunk 7]
```diff
@@ -1,110 +1,78 @@
...
+                        const command = `didc random -t '(${candidString})'`;
+                        const candidValueString = execSyncPretty(command)
+                            .toString()
+                            .trim();
+                        console.info(candidValueString);
+
+                        const canister =
+                            await getCanisterActor<Actor>('canister');
+
+                        const encodedBytes =
```
Comment (lastmjs) [inline tests/property/ic_api/candid/test/tests.ts:64]: Should we also go the other way?

**Review by lastmjs — CHANGES_REQUESTED** [review]

File: tests/property/ic_api/candid/test/tests.ts (hunk 7 above)
Comment (bdemann) [reply tests/property/ic_api/candid/test/tests.ts:64]: We do on line 67. Unless you mean start with bytes and go to candid. That is VERY difficult to do with arbitrary candid. It was pretty straight forward with just strings, but I thought we were not wanting to do that anymore

File: tests/property/ic_api/candid/test/tests.ts (hunk 5 above)
Comment (bdemann) [reply tests/property/ic_api/candid/test/tests.ts:2]: done

File: tests/property/ic_api/candid/test/tests.ts (hunk 6 above)
Comment (bdemann) [reply tests/property/ic_api/candid/test/tests.ts:28]: Done

File: tests/property/ic_api/candid/test/tests.ts (hunk 5 above)
Comment (lastmjs) [reply tests/property/ic_api/candid/test/tests.ts:2]: Let's not use execSyncPretty in our tests moving forward

File: tests/property/ic_api/candid/test/tests.ts (hunk 7 above)
Comment (lastmjs) [reply tests/property/ic_api/candid/test/tests.ts:64]: Comment about doing this once we have the ultimate power arbitrary?

File: tests/property/ic_api/candid/test/tests.ts (hunk 5 above)
Comment (bdemann) [reply tests/property/ic_api/candid/test/tests.ts:2]: done

File: tests/property/ic_api/candid/test/tests.ts (hunk 7 above)
Comment (bdemann) [reply tests/property/ic_api/candid/test/tests.ts:64]: done

**Review by lastmjs — APPROVED** [review]

**Pull request merged**